examples/tutorial.dx

```diff
@@ -0,0 +1,494 @@
+'# Dex Tutorial
+
+
+'Dex is a functional, statically typed language for array processing.
+There are many tools for array processing from high-level libraries
+like NumPy / MATLAB to low-level languages like CUDA. Dex gives you
+many benefit of the safety and simplicity benefits of high-level array
+processing languages, without requiring that you give up low-level
+control.
+
+'## Array Comprehensions
```

Comment:
Arrays as `for` comprehensions is a great way to explain `for` in Dex to folks familiar with Python!

I've been explaining Dex arrays to folks in a different way: arrays as functions. This is similar to the "Arrays are like functions" section of [the Dex workshop paper](https://openreview.net/pdf?id=rJxd7vsWPS).
- Many languages support function types. In Dex, function types look like `a -> b` and function values (lambdas) look like `\x. f x`.
- Dex arrays are like functions (from indices to elements). This is apparent in the syntax: both for array types (`n => a`) and in "array lambdas" (`for i. xs.i`). Array literal syntax (e.g. `[1, 2, 3]`) is also supported.

I wonder if calling out "array are like functions" more explicitly would make sense in a tutorial like this. This isomorphism is mentioned later in the tutorial (your comments on "array indexing is like function application").

Reply:
Yeah, I also found "arrays as functions" super appealing when I read the workshop paper. That's a great top-down way if thinking about it. 

Weirdly though when programming I find that to be a less helpful analogy. The issue is that in python I don't write extremely curried code, so I can't just convert things I think of as functions to arrays.

Typed for comprehensions though seems like a really nice constructive path? I.e. Dex is a simpler language, no magic array broadcasting, just comprehensions for everything. And they are more concise than the ones you are used to.

Reply:
I agree! "Type-inferred array comprehensions" is a nicer explanation for mainstream programmers than "arrays as functions", which is more programming-language-theoretic.

Maybe "array as functions" could be emphasized in a tutorial document or paper for a different audience, extolling the benefits of that paradigm (e.g. the compiler can choose to materialize "`for`-constructed array functions" or keep them represented as lazy indexer-functions to avoid unnecessary materialization, like the PR description for https://github.com/google-research/dex-lang/pull/399).

Reply:
That's a really good point. End programmers will definitely need to understand whether these are lazy or not, and the python analogy very much breaks down there. 

I think we should add a section discussing this. For instance doing some examples that would fail in an eager environment. (For example some the project euler examples). The question is whether it is important that they are called functions or whether "possibly lazy comprehension" is more direct.

Reply:
> I think we should add a section discussing this. For instance doing some examples that would fail in an eager environment. (For example some the project euler examples). The question is whether it is important that they are called functions or whether "possibly lazy comprehension" is more direct.

Great points.

---

An aside: I'm excited to start writing a tutorial (maybe separate, maybe merged into this one) about the Dex `List` programming model. A tentative title is "`List` manipulation in Dex". In short, Dex lists are C++-like and not Haskell-like. This is also true of the language in general: Dex compiles like C++ and avoids FP pitfalls like laziness (pointers/thunks everywhere) and unpredictable performance.

I was motivated by your [attention implementation in Dex](https://github.com/srush/learns-dex/blob/main/transformer.dx) to write an end-to-end attention notebook, including data loading. [Here's my work-in-progress!](https://github.com/dan-zheng/dex-lang/tree/attention-e2e)

I needed to implement a `splitLines : String -> List String` function, and tried writing it like Haskell (adding `cons` and `snoc` helpers for appending to Dex lists, treating Dex lists like Haskell's linked lists). This led to terrible performance: calling `splitLines` on [a 9.1 MB English → French translation dataset](https://github.com/L1aoXingyu/seq2seq-translation/blob/master/data/eng-fra.txt) basically never finished.

I talked to @apaszke and @dougalm about my poor experience. They shared that this is because Dex lists are not at all Haskell-like (lazy linked lists) and much more C++-like (contiguous arrays). Currently, it's difficult to achieve performant `List` manipulation, for functions returning `List a` (output size is unknown).

The best example of achieving performance is `def concat : (n => List a) -> List a` in the Prelude, which is `O(2n)` via doing two passes: (1) first computing the output list length and allocating an zeroed array of that length, and (2) iterating over the array to emplace elements.

https://github.com/google-research/dex-lang/blob/85ff5167e31bef734ef13552d6aad6524dc0dfc5/lib/prelude.dx#L1475-L1493

Some thoughts:
* Maybe it's possible to generalize the "design pattern" in `def concat` to a higher-order helper function for implementing `List`-returning functions. This `listConstructor` helper function would take (1) an argument function for computing the output list's length and (2) another argument function for emplacing elements into a zeroed array of that length.
* The implementation of `def concat` is quite special-cased and hardcoded. A valuable direction is generalizing the `Accum` effect (~monoidal writer monad, currently defines `+=`) to support arbitrary monoids instead of being limited to floats and float arrays: https://github.com/google-research/dex-lang/issues/298. This would unblock some `List` manipulation performance, for functions like `splitLines` that only want "list appending/concatenation" functionality. Another use case for generalized `Accum` by @duvenaud is implementing a `Stack` ADT for simulating recursion, which Dex (intentionally) does not natively support.
* There are two paths to achieving list performance: Haskell (lazy linked lists) and C++ (mutable resizable vectors). Dex lists want to be like C++ lists - implementing efficient resizable `std::vector` in Dex may greatly benefit from generalized `Accum`?

Hope this context on our recent project discussions is helpful :) <sub>(sorry if I've misrepresented anyone, feel free to chime in)</sub>

Reply:
An aside: I'm excited to start writing a tutorial (maybe separate, maybe merged into this one) about the Dex `List` programming model. A tentative title is "`List` manipulation in Dex". In short, Dex lists are C++-like and not Haskell-like. This is also true of the language in general: Dex compiles like C++ and avoids FP pitfalls like laziness (pointers/thunks everywhere) and unpredictable performance.

I was motivated by your [attention implementation in Dex](https://github.com/srush/learns-dex/blob/main/transformer.dx) to write an end-to-end attention notebook, including data loading. [Here's my work-in-progress!](https://github.com/dan-zheng/dex-lang/tree/attention-e2e)

I needed to implement a `splitLines : String -> List String` function, and tried writing it like Haskell (adding a `cons` helpers for appending elements to Dex lists, treating Dex "size-erased array" lists like Haskell's linked lists). This led to terrible asymptotic performance: calling `splitLines` on [a 9.1 MB English → French translation dataset](https://github.com/L1aoXingyu/seq2seq-translation/blob/master/data/eng-fra.txt) basically runs forever.

I asked @apaszke and @dougalm about my poor experience. They shared that this is because Dex lists are not at all designed to be Haskell-like (lazy linked lists) and much more C++-like (contiguous arrays). *This makes sense: I wasn't using Dex's intended programming model. The intended programming model is worth explaining in a tutorial!*

---

Currently, it's difficult to achieve performant `List` manipulation for functions returning `List a` (output size is unknown). The best example of achieving performance is `def concat : (n => List a) -> List a` in the Prelude, which is `O(2n)` via two passes: (1) first computing the output list length and allocating an zeroed array of that length, and (2) iterating over the array to emplace elements.

https://github.com/google-research/dex-lang/blob/85ff5167e31bef734ef13552d6aad6524dc0dfc5/lib/prelude.dx#L1475-L1493

Some thoughts:
* Maybe it's possible to generalize the "design pattern" in `def concat` to a higher-order helper function for implementing `List`-returning functions. This `listConstructor` helper function would take (1) an argument function for computing the output list's length and (2) another argument function for emplacing elements into a zeroed array of that length.
* The implementation of `def concat` is quite special-cased and hardcoded. A valuable direction is generalizing the `Accum` effect (~monoidal writer monad, currently defines `+=`) to support arbitrary monoids instead of being limited to floats and float arrays: https://github.com/google-research/dex-lang/issues/298. This would unblock some `List` manipulation performance, for functions like `splitLines` that only want "list appending/concatenation" functionality. Another use case for generalized `Accum` (from @duvenaud?) is implementing a `Stack` ADT for simulating recursion, since Dex (intentionally) does not natively support recursive functions.
* There are two paths to achieving list performance: Haskell (lazy linked lists) and C++ (mutable resizable vectors with amortized `O(1)` append). Dex lists want to be like C++ lists - implementing efficient resizable `std::vector` in Dex may greatly benefit from generalized `Accum`?

Hope this context on our recent project discussions is helpful :) <sub>(sorry if I've misrepresented anyone, feel free to chime in)</sub>

Reply:
This is extremely interesting, but certainly advanced. I would love to chat more about this and potentially help out on an attention implementation. Beyond understanding the types though, I really haven't gone deep on List yet. Sounds tricky. 

Maybe we should start a thread somewhere else?

Reply:
> I would love to chat more about this and potentially help out on an attention implementation.

That sounds fun! Maybe we could do some peer programming stream? My [e2e attention work-in-progress](https://github.com/dan-zheng/dex-lang/tree/attention-e2e) is in good shape for peer programming: next step is rewriting the string-splitting function to be more performant.

```haskell
def split (s: String) (delimiter: Char) : List String =
  ...
  -- Want an implementation with good asymptotic performance.
  -- We can try the O(n) design pattern from `def concat`.

def splitNewline (s:String) : List String =
  split s newlineChar
```

> Maybe we should start a thread somewhere else?

I think a peer programming stream would be more productive than discussion-based threads about these advanced topics, so we can communicate ideas more quickly. Also sounds more fun. What do you think? :)

Reply:
This is great @dan-zheng, but I agree with @srush that it might be a good idea to move this to a separate discussion, to avoid getting distracted from this PR!

Reply:
Thanks for the feedback! I didn't mean to dominate or distract from the PR, sorry - resolving this conversation now.